torch/csrc/api/src/nn/modules/embedding.cpp

```diff
@@ -11,27 +11,79 @@
 
-EmbeddingImpl::EmbeddingImpl(EmbeddingOptions options) : options(options) {
-  reset();
+EmbeddingImpl::EmbeddingImpl(int64_t num_embeddings, int64_t embedding_dim, c10::optional<int64_t> padding_idx, c10::optional<float> max_norm,
+float norm_type, bool scale_grad_by_freq, bool sparse, c10::optional<torch::Tensor> weight): EmbeddingImpl(EmbeddingOptions(num_embeddings, embedding_dim)){
+  if (padding_idx != c10::nullopt){
+    if(*padding_idx > 0){
+      assert((padding_idx < num_embeddings) && "Padding_idx must be within num_embeddings");
+    }
+    else{
+      assert((padding_idx >= -num_embeddings) && "Padding_idx must be within num_embedding");
+      *padding_idx = *padding_idx+num_embeddings;
+      options.padding_idx_ = padding_idx;
+    }
+  }
+  if(max_norm != c10::nullopt){
+    options.max_norm(max_norm);
+  }
+  options.norm_type(norm_type);
+  options.scale_grad_by_freq(scale_grad_by_freq);
+
+  if (!weight.has_value()){
+  // if (weight ==  c10::nullopt){
+    options.weight(torch::empty({num_embeddings, embedding_dim}));
+    EmbeddingImpl::reset();
+  }
+  else{
+    assert((((*weight).size(0) == num_embeddings) && ((*weight).size(1) == embedding_dim)) && "Shape of weight does not match num_embeddings and embedding_dim");
+  }
 }
 
 void EmbeddingImpl::reset() {
-  weight = register_parameter(
-      "weight", torch::empty({options.count_, options.dimension_}));
-  NoGradGuard guard;
-  weight.normal_(0, 1);
+  (*(options.weight_)).Tensor::normal_(0, 1);
 }
 
 void EmbeddingImpl::pretty_print(std::ostream& stream) const {
-  stream << "torch::nn::Embedding(count=" << options.count_
-         << ", dimension=" << options.dimension_ << ")";
+  stream << "torch::nn::Embedding(num_embeddings=" << options.num_embeddings()
+         << ", embedding_dim=" << options.embedding_dim();
+  if(options.padding_idx() != c10::nullopt){
+    stream << ",padding_idx=" << *options.padding_idx();
+  }
+  if(options.max_norm() != c10::nullopt){
+    stream << ",max_norm=" << *options.max_norm();
+  }
+  if(options.norm_type() != 2){
+    stream << ",norm_type=" << options.norm_type();
+  }
+  if(options.scale_grad_by_freq()){
+    stream << ",scale_grad_by_freq=" << options.scale_grad_by_freq();
+  }
+  if(options.sparse()){
+    stream << ",sparse=" << options.sparse();
+  }
+  stream << ")";
 }
 
 Tensor EmbeddingImpl::forward(const Tensor& input) {
-  return torch::embedding(weight, /*indices=*/input);
+  if(options.padding_idx() != c10::nullopt){
+    if(*options.padding_idx() > 0){
+      assert((*options.padding_idx() < (*options.weight()).size(0)) && "Padding_idx must be within num_embeddings");
+    }
+    else{
+      assert((*options.padding_idx() >= -(*options.weight()).size(0)) && "Padding_idx must be within num_embedding");
+      options.padding_idx(*options.padding_idx() + (*options.weight_).size(0));
```

Comment:
minor nit: we should likely swap the order here to match the Python implementation:
```suggestion
      options.padding_idx((*options.weight_).size(0) + *options.padding_idx());
```

Reply:
also `options.weight_` can be changed to `options._weight()`, assuming we rename `weight` to `_weight`.